apm-agent-plugins/apm-awslambda-plugin/src/main/java/co/elastic/apm/agent/awslambda/helper/AbstractAPIGatewayTransactionHelper.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.awslambda.helper;
+
+import co.elastic.apm.agent.impl.ElasticApmTracer;
+import co.elastic.apm.agent.impl.context.CloudOrigin;
+import co.elastic.apm.agent.impl.context.Request;
+import co.elastic.apm.agent.impl.context.Response;
+import co.elastic.apm.agent.impl.context.ServiceOrigin;
+import co.elastic.apm.agent.impl.context.web.ResultUtil;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.matcher.WildcardMatcher;
+import com.amazonaws.services.lambda.runtime.Context;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static co.elastic.apm.agent.configuration.CoreConfiguration.EventType.OFF;
+
+public abstract class AbstractAPIGatewayTransactionHelper<I, O> extends AbstractLambdaTransactionHelper<I, O> {
+    private static final Logger logger = LoggerFactory.getLogger(AbstractAPIGatewayTransactionHelper.class);
+    protected static final String TRANSACTION_TYPE = "request";
+    private static final String CONTENT_TYPE_HEADER = "Content-Type";
+    private static final Set<String> METHODS_WITH_BODY = new HashSet<>(Arrays.asList("POST", "PUT", "PATCH", "DELETE"));
+    private static final String CONTENT_TYPE_FROM_URLENCODED = "application/x-www-form-urlencoded";
+
+    protected AbstractAPIGatewayTransactionHelper(ElasticApmTracer tracer) {
+        super(tracer);
+    }
+
+    @Nullable
+    protected abstract String getHttpMethod(I event);
+
+    protected abstract String getApiGatewayVersion();
+
+    protected void fillHttpRequestData(Transaction transaction, @Nullable String httpMethod, @Nullable Map<String, String> headers, @Nullable String serverName, @Nullable String path, @Nullable String queryString, @Nullable String body) {
+        Request request = transaction.getContext().getRequest();
+        if (null != httpMethod) {
```

Comment:
Why do we give up all data when `httpMethod == null`?

Reply:
You are right we can extract 

```
            fillUrlRelatedFields(request, serverName, path, queryString);
```

from the if block. The rest (i.e. body capturing) depends on the http method.